infer parquet file order from metadata and use it to optimize scans

adriangb opened the pull request:
The idea here is to use the metadata in parquet files to infer sort orders, thus it is not required for users to specify it manually.

This should probably be split into multiple PRs:
- Record sort order when writing into a table created as `WITH ORDER`
- Refactor `PartitionedFile` construction
- Collect ordering during statistics collection

zhuqi-lucas commented on datafusion/sqllogictest/test_files/parquet.slt:
```diff
@@ -885,3 +885,42 @@ WHERE b = 2;
 
 statement ok
 DROP TABLE t;
+
```
It seems only test ASC NULLS FIRST. Better to add test cases for:
     - DESC ordering
     - Multi-column ordering
     - Files with no ordering metadata

zhuqi-lucas commented on datafusion/physical-expr-common/src/sort_expr.rs:
```diff
@@ -458,6 +458,21 @@ impl LexOrdering {
     }
 
+    /// Checks if `other` is a prefix of this `LexOrdering`.
+    pub fn is_prefix(&self, other: &LexOrdering) -> bool {
```
I meet some similar topic for the sort pushdown, and submitted a [PR](https://github.com/apache/datafusion/pull/19446#issue-3752083820), i prefer to use eq_properties to do similar things, it will handle more cases after my try in my internal project. 


The ordering_satisfy will handle more cases than prefix.

adriangb replied:
I gave it a review - looks nice! I can sequence this behind your PR if needed.

adriangb commented on datafusion/core/tests/physical_optimizer/filter_pushdown/mod.rs:
```diff
@@ -1829,7 +1829,7 @@ async fn test_topk_dynamic_filter_pushdown_integration() {
   SELECT 1372708800 + value AS t
   FROM generate_series(0, 99999)
-  ORDER BY t
+  ORDER BY t + 1
```
Because it was picking up the ordering and causing even less data to be read 😆

zhuqi-lucas commented on datafusion/catalog-listing/src/table.rs:
```diff
@@ -336,17 +336,64 @@ impl ListingTable {
+            if let Some(ordering) = &file.ordering {
+                all_orderings.push(ordering);
+            } else {
+                // If any file has no ordering, we can't derive a common ordering
```
We may add some debug or tracing log here:

```suggestion
                // If any file has no ordering, we can't derive a common ordering
                tracing!(
        "Cannot derive common ordering: file {} has different ordering ({:?}) than first file ({:?})",
        file.object_meta.location, ordering, first
    );
```

zhuqi-lucas commented on datafusion/datasource-parquet/src/file_format.rs:
```diff
@@ -500,7 +542,22 @@ impl FileFormat for ParquetFormat {
+            // may not be compatible with Parquet sorting columns (e.g. ordering on `random()`).
+            // So if we cannot create a Parquet sorting column from the ordering requirement,
+            // we skip setting sorting columns on the Parquet sink.
+            lex_ordering_to_sorting_columns(&ordering).ok()
```
Great point!

alamb reviewed (commented):
Thanks @zhuqi-lucas  and @adriangb  -- I took a quick skim of this and it looks like a great idea. Let me know if there is anything else that needs a review -- otherwise 🚀

adriangb commented:
I've split out a couple of PRs from this:
- https://github.com/apache/datafusion/pull/19595
- https://github.com/apache/datafusion/pull/19596
- https://github.com/apache/datafusion/pull/19597

adriangb commented:
@zhuqi-lucas I think this is ready for another review now, I've incorporated the changes from the broken out PRs, rebased on main and addressed your previous feedback.

zhuqi-lucas reviewed (approved):
Great work @adriangb, LGTM now!
